### DISCUSSION TITLE
Data Structures across shadow and OS context

### DISCUSSION THREAD

**Discussion by akumar1512** [shadow/shadow#2836] [Q&A] [answered]
This is a continuation to the [previous discussion](https://github.com/shadow/shadow/discussions/2786#discussioncomment-5315105) related to usage of shadow for larger federated networks.

I have a piece of code in the management app (app that runs alongside simulating app and takes care to handle out of shadow calls) that receives socket (socket that listens to requests made by simulating app inside shadow) and makes out of shadow calls on external socket on real OS.

```
self.map_sockets = {}
def process_socket(self, shadow_sim_app_socket):
    sim_request = shadow_sim_app_socket.recv(4096).decode('utf-8')
    print("[INSIDE SHADOW] Simulation app <-> management app socket - {}".format(shadow_sim_app_socket))
    
    ShadowInterposition.disable()
    
    external_server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    external_server_socket.connect((SHIMHOST, SHIMPORT))
    external_server_socket.sendall(silo_request.encode('utf-8'))
    self.map_sockets[external_server_socket] = shadow_sim_app_socket

    print("[INSIDE OS] Map content - {}".format(self.map_sockets))
    ShadowInterposition.enable()
    
    print("[INSIDE SHADOW] Map content - {}".format(self.map_sockets))
    
    
Output:
[INSIDE SHADOW]  Simulation app <-> management app socket - <ssl.SSLSocket fd=5, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 443), raddr=('11.0.0.2', 30030)>

# shadow_sim_app_socket shows a different value (expected)
[INSIDE OS] Map content -  {<socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 43586), raddr=('127.0.0.1', 8998)>: <ssl.SSLSocket fd=5, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 8999)>}

# Similarly external_server_socket shows different value 
[INSIDE SHADOW] Map content -  {<socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>: <ssl.SSLSocket fd=5, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 443), raddr=('11.0.0.2', 30030)>}
```

I have been using enabling and disabling interposition carefully till now and it has served most of my use-cases. Even blocking implementation works just fine. Going ahead I see a need to have certain data structures that is shared between both contexts. As for in the example above I need to maintain this map of sockets to save sockets and enable non-blocking calls using select.

I understand this is an expected behavior but is there anyway I can define data structures that I can use across shadow and OS? Thanks for your help.

**Comment by stevenengler** [comment, marked as answer]
What do you mean by "define data structures that I can use across shadow and OS"? Your dictionary data structure is valid across both contexts (interposition enabled and disabled), but the sockets that it stores are not valid across both contexts. More specifically, the sockets you create in one context are not valid in the other context. In your example, the keys of the dictionary are only valid when interposition is disabled, and the values are only valid when interposition is enabled.

> As for in the example above I need to maintain this map of sockets to save sockets and enable non-blocking calls using select.

I'm not quite sure what you mean here. If interposition is enabled, you can only `select` on fds that were created when interposition was enabled. If interposition is disabled, you can only `select` on fds that were created when interposition was disabled. You can't mix them together.

**Reply by akumar1512** [reply]
Thanks @stevenengler for your response. I was actually getting confused by the difference in debug prints of same sockets in both context. Just saw the hash and fileno, which is same in both contexts.

> If interposition is enabled, you can only select on fds that were created when interposition was enabled

Actually I think the issue was here. Thanks for this info, I will keep a note to this for my management app design.

**Reply by stevenengler** [reply]
The Python socket object will store the file handle (fd) and probably some other information about the socket like the socket family. When you debug print the socket, Python will probably make syscalls to do things like get the socket address, peer name, etc to display it (the laddr and raddr properties in your output). If the socket was created when interposition is enabled, then you disable interposition and debug print the socket, Python will perform these syscalls on the same integer file handle but the file handle won't refer to the original socket (and there may not even be a valid socket for that file handle). So you'll print the debug information for an entirely different (or no) socket, and you may cause Python to get confused.

There's a list of some other things that aren't valid across contexts in the comment here: https://github.com/shadow/shadow/issues/2157#issue-1237693886

> Here the file will be opened directly in the plugin process. When native syscalls are enabled, plugin code would not be able to access things like fds (including stdout/stderr), mutexes/futexes, etc that were created when running under Shadow. You would probably also want to make sure not to invoke signals, and I'm sure there are a lot of other things that could go wrong if you're not careful.

**Edit:** Here's how python's socket object is defined: https://github.com/python/cpython/blob/08e4e887f2d4650972272b2f4441485f3e1e9aab/Modules/socketmodule.h#L314
